ext/standard/uuid.c

```diff
@@ -0,0 +1,662 @@
+{
+	PHP_MD5_CTX context;
+	unsigned char digest[16];
+
+	PHP_MD5Init(&context);
+	PHP_MD5Update(&context, namespace->bytes, PHP_UUID_LEN);
+	PHP_MD5Update(&context, name, name_len);
+	PHP_MD5Final(digest, &context);
+	memcpy(uuid->bytes, digest, PHP_UUID_LEN);
+	set_variant_rfc4122(uuid);
+	php_uuid_set_version(uuid, PHP_UUID_VERSION_3_NAME_BASED_MD5);
+}
+
+PHPAPI int php_uuid_create_v4(php_uuid *uuid, const zend_bool throw)
+{
+	int result = php_random_bytes(uuid, PHP_UUID_LEN, throw);
+
+	if (result == SUCCESS) {
+		set_variant_rfc4122(uuid);
+		php_uuid_set_version(uuid, PHP_UUID_VERSION_4_RANDOM);
+	}
+
+	return result;
+}
+
+PHPAPI void php_uuid_create_v5(php_uuid *uuid, const php_uuid *namespace, const char *name, const size_t name_len)
+{
+	PHP_SHA1_CTX context;
+	unsigned char digest[20];
+
+	PHP_SHA1Init(&context);
+	PHP_SHA1Update(&context, (const unsigned char *) namespace->bytes, PHP_UUID_LEN);
+	PHP_SHA1Update(&context, (const unsigned char *)name, name_len);
+	PHP_SHA1Final(digest, &context);
+	memcpy(uuid->bytes, digest, PHP_UUID_LEN);
+	set_variant_rfc4122(uuid);
+	php_uuid_set_version(uuid, PHP_UUID_VERSION_5_NAME_BASED_SHA1);
+}
+
+static ZEND_COLD zend_object *throw_uuid_parse_exception(const char *input, const size_t input_len, const size_t position, const char *reason, ...)
+{
+	va_list format_args;
+	char *formatted_reason;
+	zend_object *object;
+	zval exception;
+
+	va_start(format_args, reason);
+	zend_vspprintf(&formatted_reason, 0, reason, format_args);
+	va_end(format_args);
+	object = zend_throw_exception(php_ce_UUIDParseException, formatted_reason, 0);
+	efree(formatted_reason);
+
+	ZVAL_OBJ(&exception, object);
+	zend_update_property_stringl(php_ce_UUIDParseException, &exception, UUID_EX_INPUT_PROP, UUID_EX_INPUT_PROP_LEN, input, input_len);
+	zend_update_property_long(php_ce_UUIDParseException, &exception, UUID_EX_POSITION_PROP, UUID_EX_POSITON_PROP_LEN, position);
+
+	return object;
+}
+
+static zend_always_inline zend_object *throw_uuid_parse_exception_invalid_len(const char *input, const size_t input_len, const size_t position, const size_t actual)
+{
+	return throw_uuid_parse_exception(
+		input,
+		input_len,
+		position,
+		"Expected at least %u hexadecimal digits, but got %u",
+		UUID_HEX_LEN,
+		actual
+	);
+}
+
+static zend_always_inline zend_object *throw_uuid_parse_exception_invalid_char(const char *input, const size_t input_len, const size_t position)
+{
+	return throw_uuid_parse_exception(
+		input,
+		input_len,
+		position,
+		"Expected hexadecimal digit, but found '%c' (0x%02x)",
+		input[position],
+		input[position]
+	);
+}
+
+PHPAPI int php_uuid_parse(php_uuid *uuid, const char *input, const size_t input_len, const zend_bool throw)
+{
+	size_t position = 0;
+	size_t digit    = 0;
+	size_t limit    = input_len - 1;
+	uint8_t byte    = 0;
+
+	while (input[position] == '-' || input[position] == ' ' || input[position] == '\t' || input[position] == '{') {
```

Comment:
Where do these characters come from?

Reply:
- ` ` and `\t` are non-harmful possible leading and trailing characters from user input
- `-` might come from a non-compliant UUID genrator
- `{` and `}` is from the Microsoft string representation